--- a/wan/modules/causal_model.py
+++ b/wan/modules/causal_model.py
@@ -170,7 +170,7 @@
                     key=padded_roped_key.transpose(2, 1),
                     value=padded_v.transpose(2, 1),
                     block_mask=block_mask
-                )[:, :, :-padded_length].transpose(2, 1)
+                )[:, :, :q.shape[1]].transpose(2, 1)
 
             else:
                 roped_query = rope_apply(q, grid_sizes, freqs).type_as(v)
@@ -201,7 +201,7 @@
                     key=padded_roped_key.transpose(2, 1),
                     value=padded_v.transpose(2, 1),
                     block_mask=block_mask
-                )[:, :, :-padded_length].transpose(2, 1)
+                )[:, :, :q.shape[1]].transpose(2, 1)
         else:
             frame_seqlen = math.prod(grid_sizes[0][1:]).item()
             current_start_frame = current_start // frame_seqlen
